packages/kbn-config-schema/src/helpers/on_serverless.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '../..';
+import { Type, TypeOptions } from '../types';
+
+/**
+ * Helper to apply different validations depending on whether Kibana is running on Serverless or not.
+ *
+ * @example Only allow the setting on Serverless
+ * const config = schema.object({
+ *   myProp: onServerless(schema.boolean({ defaultValue: true })),
+ * });
+ *
+ * @example Fixed value on self-managed, configurable on Serverless
+ * const config = schema.object({
+ *   myProp: onServerless(schema.boolean({ defaultValue: true }), schema.literal(false), {
+ *     defaultValue: false,
+ *   }),
+ * });
```

Comment:
I gotta say, seeing `{ default: false }` and trying to reason about what this config does makes my head spin a little bit 😅  -- I know this is more from how `schema.conditional` works.

Does make me wonder, for this functionality in what cases do we need that extra `options` setting?

Reply:
When using `schema.literal()`, you need to handle the possibility of not having a value.

So, either you do `schema.maybe(schema.literal(false))`, or you set the default value at the end. Neither `.literal`, nor `.maybe` allow setting the defaultValue in their constructors, so you need to rely on the conditional options to do that.

In any case, following your suggestion of fluent API, we could do something like

```ts
schema.object({
   myProp: onServerless(schema.boolean()).else(schema.literal(false)).default(false)
 })
```

WDYT?

Reply:
On the flip side, the "magic" of having the `{ defaultValue }` object is that it also allows us to use other options like `validate`, where custom messages can be returned if necessary.